security/pkg/nodeagent/caclient/providers/vault1/client_test.go

```diff
@@ -0,0 +1,350 @@
+		}
+
+		resp, err := cli.CSRSign(context.Background(), tc.cliConfig.csr, tc.cliConfig.clientToken, 1)
+		if err != nil {
+			match, _ := regexp.MatchString(tc.expectedErr+".+", err.Error())
+			if !match {
+				t.Errorf("Test case [%s]: error (%s) does not match expected error (%s)", id, err.Error(), tc.expectedErr)
+			}
+		} else {
+			if tc.expectedErr != "" {
+				t.Errorf("Test case [%s]: expect error: %s but got no error", id, tc.expectedErr)
+			} else if !reflect.DeepEqual(resp, tc.expectedCert) {
+				t.Errorf("Test case [%s]: resp: got %+v, expected %v", id, resp, tc.expectedCert)
+			}
+		}
+	}
+}
+
+func TestClientOnExampleHttpVaultCA(t *testing.T) {
+	testCases := map[string]struct {
+		cliConfig clientConfig
+	}{
+		"Valid certs 1": {
+			cliConfig: clientConfig{vaultAddr: vaultNonTLSAddr, vaultLoginPath: "auth/kubernetes/login", vaultLoginRole: "istio-cert", vaultSignCsrPath: "istio_ca/sign/istio-pki-role", clientToken: citadelSANonTLS, csr: []byte(testCsr1)},
+		},
+	}
+
+	for id, tc := range testCases {
+		var vaultAddr string
+		vaultAddr = tc.cliConfig.vaultAddr
+		cli, err := NewVaultClient1(false, []byte{}, vaultAddr, tc.cliConfig.vaultLoginRole,
+			tc.cliConfig.vaultLoginPath, tc.cliConfig.vaultSignCsrPath)
+		if err != nil {
+			t.Errorf("Test case [%s]: failed to create ca client: %v", id, err)
+		}
+
+		resp, err := cli.CSRSign(context.Background(), tc.cliConfig.csr, tc.cliConfig.clientToken, 1)
+		if err != nil {
+			t.Errorf("Test case [%s]:  error (%v) is not expected", id, err.Error())
+		} else {
+			if len(resp) != 3 {
+				t.Errorf("Test case [%s]: the certificate chain length (%v) is unexpected", id, len(resp))
+			}
+		}
+	}
+}
+
+func TestClientOnExampleHttpsVaultCA(t *testing.T) {
+	testCases := map[string]struct {
+		cliConfig clientConfig
+	}{
+		"Valid certs 1": {
+			cliConfig: clientConfig{vaultAddr: vaultTLSAddr, vaultLoginPath: "auth/kubernetes/login", vaultLoginRole: "istio-cert", vaultSignCsrPath: "istio_ca/sign/istio-pki-role", clientToken: citadelSATLS, csr: []byte(testCsr1)},
+		},
+	}
+
+	for id, tc := range testCases {
+		var vaultAddr string
+		vaultAddr = tc.cliConfig.vaultAddr
+		cli, err := NewVaultClient1(true, []byte(vaultServerTlsCert), vaultAddr, tc.cliConfig.vaultLoginRole,
+			tc.cliConfig.vaultLoginPath, tc.cliConfig.vaultSignCsrPath)
+		if err != nil {
+			t.Errorf("Test case [%s]: failed to create ca client: %v", id, err)
+		}
+
+		resp, err := cli.CSRSign(context.Background(), tc.cliConfig.csr, tc.cliConfig.clientToken, 1)
+		if err != nil {
+			t.Errorf("Test case [%s]:  error (%v) is not expected", id, err.Error())
+		} else {
+			if len(resp) != 3 {
+				t.Errorf("Test case [%s]: the certificate chain length (%v) is unexpected", id, len(resp))
+			}
+		}
+	}
+}
+
+// NewMockVaultServer creates a mock Vault server for testing purpose.
+// token: required access token
+func NewMockVaultServer(t *testing.T, tls bool, loginRole, token, loginResp, signResp string) *MockVaultServer {
```

Comment:
Make this internal by starting with lowercase letter?

Reply:
Done.